### PR TITLE
Fix schedule task memory leak in ManagedLedgerFactoryImpl

### PR DESCRIPTION
**Pull request by Shoothzj**
### Motivation
The memory is leak due to the ManagerLedgerImpl can't be garbage collected.

### Modifications

Before remove the `managedLedgerImpl` from map, call  `close()` to cacel the schdule job first.

### Documentation

Check the box below and label this PR (if you have committer privilege).

Need to update docs? 
  
- [x] `no-need-doc` 
  
 Fix a memory leak bug

### CODE REVIEW

**Comment by lhotari** [comment]
> The memory is leak due to the ManagerLedgerImpl can't be garbage collected.

Can you share more details about the symptoms? Which Pulsar versions are impacted?

**Comment by Shoothzj** [comment]
> 
The `managedledgerImpl` were removed from map, but the schedule tasks didn't stop, So their memory can't be released, leading to a memory leak.

In current branch we maintained, version 2.8 and version 2.9 are impacted.

**Comment by Shoothzj** [comment]
@codelipenghui @hangc0276 @eolivelli PTAL

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerFactoryImpl.java [hunk 1]
```diff
@@ -352,6 +352,7 @@ public void asyncOpen(final String name, final ManagedLedgerConfig config, final
                         // Managed ledger is in unusable state. Recreate it.
                         log.warn("[{}] Attempted to open ledger in {} state. Removing from the map to recreate it",
                                 name, l.getState());
+                        l.close();
```
Comment (lhotari) [inline managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerFactoryImpl.java:355]: Should it be called asynchronously after the removal from the map? What if close blocks or throws an exception?

Comment (Shoothzj) [reply managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerFactoryImpl.java]: @lhotari please take a look again

**Review by lhotari — APPROVED** [review]
LGTM

**Comment by lhotari** [comment]
Good catch @Shoothzj

**Review by merlimat — APPROVED** [review]

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerFactoryImpl.java [hunk 2]
```diff
@@ -353,6 +353,17 @@ public void asyncOpen(final String name, final ManagedLedgerConfig config, final
                         log.warn("[{}] Attempted to open ledger in {} state. Removing from the map to recreate it",
                                 name, l.getState());
                         ledgers.remove(name, existingFuture);
+                        l.asyncClose(new CloseCallback() {
```
Comment (hangc0276) [inline managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerFactoryImpl.java:356]: Actually, if the managedLedger in `Fenced` or `Closed` state, call `asyncClose` just
https://github.com/apache/pulsar/blob/ee904063abc8ec90c3c536b28a47a59975175cd4/managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java#L1316-L1328

So, should we just call `ManagedLedgerFactoryImpl::close` ?

Comment (lhotari) [reply managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerFactoryImpl.java:356]: Good point. I guess calling asyncClose is sufficient. However there  seems to be a bug in the asyncClose method when the ledger is fenced. The method call to cancel the scheduled tasks is missing.

Comment (lhotari) [reply managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerFactoryImpl.java:356]: +1

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java [hunk 3]
```diff
@@ -1192,6 +1192,7 @@ private long consumedLedgerSize(long ledgerSize, long ledgerEntries, long consum
     @Override
     public synchronized void asyncTerminate(TerminateCallback callback, Object ctx) {
         if (state == State.Fenced) {
+            cancelScheduledTasks();
```
Comment (lhotari) [inline managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java:1195]: +1

**Review by lhotari — APPROVED** [review]

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java (hunk 3 above)
Comment (lhotari) [reply managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java:1195]: It seems that similar addition would be needed to asyncClose when the ledger is fenced

Comment (Shoothzj) [reply managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java:1195]: Yes, yesterday，I found It has been fixed in #12565, it's better to cancel the schedule task when the ledger is fenced. I will test if it fix the memomry leak.

**Pull request closed without merging**